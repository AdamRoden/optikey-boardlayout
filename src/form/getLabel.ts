const getLabel = (label: string) => {
  const i = label.indexOf("]")
  return i === -1 ? label : label.substring(i + 2)
}

export default getLabel
