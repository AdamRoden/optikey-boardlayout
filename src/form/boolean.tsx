import Checkbox from '@material-ui/core/Checkbox'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormGroup from '@material-ui/core/FormGroup'
import CheckBoxIcon from '@material-ui/icons/CheckBox'
import { getMapToFromProps, IFieldProps, JSONSchemaDefinition, memo, useFieldWithValidation } from "@xpfw/form"
import { observer } from "mobx-react-lite"
import * as React from "react"
import getLabel from "./getLabel"

const BooleanField: React.FunctionComponent<IFieldProps & {
  className?: string
  placeholder?: string
}> = observer((props) => {
  const mapTo = getMapToFromProps(props)
  const fieldHelper = useFieldWithValidation(props.schema, mapTo, props.prefix)
  return (
    <FormControlLabel
        control={
          <Checkbox
            checked={fieldHelper.value}
            onChange={(e) => {
              fieldHelper.setValue(!fieldHelper.value)
            }}
            value={fieldHelper.value}
          />
        }
        label={getLabel(mapTo)}
    />
  )
})

export default BooleanField
