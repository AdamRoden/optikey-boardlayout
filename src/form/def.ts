import { ComponentRegistry, ExtendedJSONSchema } from "@xpfw/form"
import { registerComponents } from "@xpfw/form-web"
import BooleanField from "./boolean"
import SelectField from "./select"
import TextField from "./text"

registerComponents()
ComponentRegistry.registerComponent("number", TextField)
ComponentRegistry.registerComponent("string", TextField)
ComponentRegistry.registerComponent("boolean", BooleanField)
ComponentRegistry.registerComponent("string", SelectField, "select")

const nameSchema: ExtendedJSONSchema = {
  title: "Name", type: "string"
}

const heightSchema: ExtendedJSONSchema = {
  title: "Height", type: "number"
}

const symbolSchema: ExtendedJSONSchema = {
  title: "SymbolMargin", type: "number"
}

const borderSchema: ExtendedJSONSchema = {
  title: "BorderThickness", type: "string"
}

const rowsSchema: ExtendedJSONSchema = {
  title: "Grid.Rows", type: "number"
}

const colsSchema: ExtendedJSONSchema = {
  title: "Grid.Cols", type: "number"
}

const keyboardPrefix = "Keyboard"
const hoverKey = `hover`
const dragKey = `drag`
const activeKey = `active`

export {
  nameSchema, heightSchema, borderSchema, symbolSchema, rowsSchema, colsSchema, keyboardPrefix,
  activeKey, dragKey, hoverKey
}
