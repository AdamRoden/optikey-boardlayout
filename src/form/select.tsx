import FormControl from '@material-ui/core/FormControl'
import FormGroup from '@material-ui/core/FormGroup'
import FormHelperText from '@material-ui/core/FormHelperText'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import CheckBoxIcon from '@material-ui/icons/CheckBox'
import { getMapToFromProps, IFieldProps, JSONSchemaDefinition, memo, useFieldWithValidation } from "@xpfw/form"
import { get, isFunction } from "lodash"
import { observer } from "mobx-react-lite"
import * as React from "react"
import getLabel from "./getLabel";

const SelectField: React.FunctionComponent<IFieldProps & {
  className?: string
  placeholder?: string
}> = observer((props) => {
  const mapTo = getMapToFromProps(props)
  const fieldHelper = useFieldWithValidation(props.schema, mapTo, props.prefix)
  let selOpts = get(props, "schema.selectOptions", [])
  if (isFunction(selOpts)) {
    selOpts = selOpts(fieldHelper.value, props.schema, props)
  }
  const options = selOpts.map((option: any) => {
    return (
      <MenuItem key={option.value} value={option.value}>
        {option.label}
      </MenuItem>
    )
  })
  return (
    <FormControl fullWidth>
      <InputLabel>{getLabel(mapTo)}</InputLabel>
        <Select
          onChange={(e) => {
            const t: any = get(e, "nativeEvent.target")
            fieldHelper.setValue(t.getAttribute("data-value"))
          }}
          value={fieldHelper.value}
        >
          {options}
        </Select>
     </FormControl>
  )
})

export default SelectField
