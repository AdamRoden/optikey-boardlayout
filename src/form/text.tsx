import TextFiel from '@material-ui/core/TextField'
import { getMapTo, getMapToFromProps, IFieldProps, JSONSchemaDefinition, memo, useFieldWithValidation } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react-lite"
import * as momentA from "moment"
import * as React from "react"
const MTextField: any = TextFiel
const TextField: React.FunctionComponent<IFieldProps & {
  className?: string
  placeholder?: string
}> = observer((props) => {
  const format = get(props, "schema.format")
  const isDate = format === "date" || format === "date-time" || format === "time"
  const fieldHelper = useFieldWithValidation(props.schema, getMapToFromProps(props), props.prefix, {
    valueEventKey: "nativeEvent.target.value"
  })
  const fieldType = get(props, "schema.type")
  const value = fieldHelper.value
  let type = "text"
  let min
  let max
  let step
  const onChange = fieldHelper.setValue
  if (fieldType === "number") {
    type = "number"
    min = get(props, "schema.minimum")
    max = get(props, "schema.maximum")
    step = get(props, "schema.step")
  }
  return (
    <MTextField
      type={type}
      label={getMapTo(props.schema, props.mapTo)}
      id={get(props, "id")}
      className={get(props, "className")}
      value={value}
      inputProps={{step, min, max}}
      onChange={onChange}
      placeholder={props.placeholder}
      fullWidth
    />
  )
})

export default TextField
