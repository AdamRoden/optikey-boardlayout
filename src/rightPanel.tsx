import { ComponentRegistry, ExtendedJSONSchema, FormStore, SharedField } from "@xpfw/form"
import { registerComponents, TextField } from "@xpfw/form-web"
import * as React from "react"

registerComponents()
ComponentRegistry.registerComponent("number", TextField)

const nameSchema: ExtendedJSONSchema = {
  title: "Name", type: "string"
}

const heightSchema: ExtendedJSONSchema = {
  title: "Height", type: "number"
}
const symbolSchema: ExtendedJSONSchema = {
  title: "SymbolMargin", type: "number"
}
const borderSchema: ExtendedJSONSchema = {
  title: "BorderThickness", type: "string"
}
const rowsSchema: ExtendedJSONSchema = {
  title: "Grid.Rows", type: "number"
}
const colsSchema: ExtendedJSONSchema = {
  title: "Grid.Cols", type: "number"
}
const keyboardPrefix = "Keyboard"

const RightPanel: React.FunctionComponent<any> = () => (
  <div>
    hi panep
    <SharedField schema={nameSchema} prefix={keyboardPrefix} />
    <SharedField schema={heightSchema} prefix={keyboardPrefix} />
    <SharedField schema={borderSchema} prefix={keyboardPrefix} />
    <SharedField schema={rowsSchema} prefix={keyboardPrefix} />
    <SharedField schema={colsSchema} prefix={keyboardPrefix} />
    <SharedField schema={symbolSchema} prefix={keyboardPrefix} />
  </div>
)

export default RightPanel
