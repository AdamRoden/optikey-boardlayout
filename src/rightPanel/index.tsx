import Button from "@material-ui/core/Button"
import ExpansionPanel from "@material-ui/core/ExpansionPanel"
import ExpansionPanelDetails from "@material-ui/core/ExpansionPanelDetails"
import ExpansionPanelSummary from "@material-ui/core/ExpansionPanelSummary"
import Typography from "@material-ui/core/Typography"
import ExpandMoreIcon from "@material-ui/icons/ExpandMore"
import { SharedField } from "@xpfw/form"
import * as React from "react"
import {
  borderSchema, colsSchema, heightSchema, keyboardPrefix, nameSchema, rowsSchema, symbolSchema
} from "../form/def"
import downloadKeyboard, { copyToClipboard } from "../util/download"
import AboutPanel from "./about"
import KeyboardImporter from "./import"
import KeyEditor from "./keyEditor"

const RightPanel: React.FunctionComponent<any> = () => (
  <div className="rightPanel">
    <div className=" row">
      <Button variant="outlined" color="primary" fullWidth onClick={downloadKeyboard}>download xml</Button>
      <Button variant="outlined" color="primary" fullWidth onClick={copyToClipboard}>copy xml to clipboard</Button>
    </div>
    <KeyboardImporter />
    <ExpansionPanel defaultExpanded={false}>
      <ExpansionPanelSummary  expandIcon={<ExpandMoreIcon />}>
        <Typography>Global Settings</Typography>
      </ExpansionPanelSummary>
      <ExpansionPanelDetails>
        <div className="center" style={{flexDirection: "column", flex: 1}}>
          <SharedField schema={nameSchema} prefix={keyboardPrefix} />
          <SharedField schema={heightSchema} prefix={keyboardPrefix} />
          <SharedField schema={borderSchema} prefix={keyboardPrefix} />
          <SharedField schema={rowsSchema} prefix={keyboardPrefix} />
          <SharedField schema={colsSchema} prefix={keyboardPrefix} />
          <SharedField schema={symbolSchema} prefix={keyboardPrefix} />
        </div>
      </ExpansionPanelDetails>
    </ExpansionPanel>
    <KeyEditor />
    <AboutPanel />
  </div>
)

export default RightPanel
