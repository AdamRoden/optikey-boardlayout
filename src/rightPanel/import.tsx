import ExpansionPanel from '@material-ui/core/ExpansionPanel'
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails'
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary'
import Typography from '@material-ui/core/Typography'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'
import { FormStore } from "@xpfw/form"
import * as React from "react"
import { useCallback } from "react"
import { useDropzone } from "react-dropzone"
import { keyboardPrefix } from "../form/def"
import xmlParser from "../xml/parse"

const readFiles = (acceptedFiles: any) => {
  if (acceptedFiles.length > 0) {
    const reader = new FileReader()
    reader.readAsText(acceptedFiles[0])
    reader.onloadend = () => {
      const res: any = reader.result
      FormStore.setValue(keyboardPrefix, xmlParser(res).Keyboard)
    }
  }
}

const KeyboardImporter = () => {
  const onDrop = useCallback(readFiles, [])
  const {getRootProps, getInputProps, isDragActive} = useDropzone({onDrop})
  const rp: any = getRootProps()
  return (

    <ExpansionPanel {...rp} defaultExpanded={false} expanded={false}>
      <ExpansionPanelSummary>
        <input {...getInputProps()} />
        <Typography>Upload</Typography>
        <Typography>Drag 'n' drop some files here, or click to select files</Typography>
      </ExpansionPanelSummary>
    </ExpansionPanel>
  )
}

export default KeyboardImporter
export {
  readFiles
}
