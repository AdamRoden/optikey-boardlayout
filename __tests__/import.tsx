import { toJS } from "@xpfw/data"
import { FormStore } from "@xpfw/form"
import { keyboardPrefix } from "../src/form/def"
import { readFiles } from "../src/rightPanel/import"
import exampleXml from "../src/xml/example"

test("READ FILES", (done) => {
  expect(toJS(FormStore.getValue(keyboardPrefix))).toMatchSnapshot("no keyboard imported yet")
  readFiles([])
  expect(toJS(FormStore.getValue(keyboardPrefix))).toMatchSnapshot("empty import makes no changes")
  const blob = new Blob([exampleXml], {type : "application/json"})
  readFiles([blob])
  setTimeout(() => {
    expect(toJS(FormStore.getValue(keyboardPrefix))).toMatchSnapshot("imported via file")
    done()
  }, 2000)
})
