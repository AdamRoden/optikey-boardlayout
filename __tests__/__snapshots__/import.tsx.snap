// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`READ FILES: empty import makes no changes 1`] = `undefined`;

exports[`READ FILES: imported via file 1`] = `
Object {
  "BorderThickness": "0,0,0,0",
  "Grid": Object {
    "Cols": 12,
    "Rows": 7,
  },
  "Height": 40,
  "Keys": Array [
    Object {
      "Col": 8,
      "Label": "7",
      "Row": 0,
      "Text": "7",
      "type": "TextKey",
    },
    Object {
      "Col": 9,
      "Label": "8",
      "Row": 0,
      "Text": "8",
      "type": "TextKey",
    },
    Object {
      "Col": 10,
      "Label": "9",
      "Row": 0,
      "Text": "9",
      "type": "TextKey",
    },
    Object {
      "Col": 1,
      "Row": 2,
      "type": "TextKey",
    },
    Object {
      "Col": 0,
      "Label": "~",
      "Row": 0,
      "Text": "~",
      "type": "TextKey",
    },
    Object {
      "Col": 2,
      "Label": "$",
      "Row": 1,
      "Text": "$",
      "type": "TextKey",
    },
    Object {
      "Col": 1,
      "Label": "&",
      "Row": 1,
      "Text": "&",
      "type": "TextKey",
    },
    Object {
      "Col": 1,
      "Label": "^",
      "Row": 3,
      "Text": "^",
      "type": "TextKey",
    },
    Object {
      "Col": 0,
      "Label": "!",
      "Row": 2,
      "Text": "!",
      "type": "TextKey",
    },
    Object {
      "Col": 2,
      "Label": "?",
      "Row": 3,
      "Text": "?",
      "type": "TextKey",
    },
    Object {
      "Col": 10,
      "Label": "|",
      "Row": 4,
      "Text": "|",
      "type": "TextKey",
    },
    Object {
      "Action": "MenuKeyboard",
      "Col": 11,
      "Row": 0,
      "Symbol": "MenuIcon",
      "type": "ActionKey",
    },
    Object {
      "Col": 5,
      "Label": "4",
      "Row": 0,
      "Text": "4",
      "type": "TextKey",
    },
    Object {
      "Col": 6,
      "Label": "5",
      "Row": 0,
      "Text": "5",
      "type": "TextKey",
    },
    Object {
      "Col": 7,
      "Label": "6",
      "Row": 0,
      "Text": "6",
      "type": "TextKey",
    },
    Object {
      "Col": 3,
      "Label": "/",
      "Row": 1,
      "Text": "/",
      "type": "TextKey",
    },
    Object {
      "Col": 0,
      "Label": ":",
      "Row": 4,
      "Text": ":",
      "type": "TextKey",
    },
    Object {
      "Col": 6,
      "Row": 3,
      "ShiftDownLabel": "F",
      "ShiftUpLabel": "f",
      "Text": "f",
      "type": "TextKey",
    },
    Object {
      "Col": 7,
      "Row": 3,
      "ShiftDownLabel": "G",
      "ShiftUpLabel": "g",
      "Text": "g",
      "type": "TextKey",
    },
    Object {
      "Col": 8,
      "Row": 3,
      "ShiftDownLabel": "H",
      "ShiftUpLabel": "h",
      "Text": "h",
      "type": "TextKey",
    },
    Object {
      "Col": 2,
      "Row": 2,
      "ShiftDownLabel": "Q",
      "ShiftUpLabel": "q",
      "Text": "q",
      "type": "TextKey",
    },
    Object {
      "Col": 9,
      "Row": 3,
      "ShiftDownLabel": "J",
      "ShiftUpLabel": "j",
      "Text": "j",
      "type": "TextKey",
    },
    Object {
      "Col": 8,
      "Label": "%",
      "Row": 1,
      "Text": "%",
      "type": "TextKey",
    },
    Object {
      "Action": "BackOne",
      "Col": 8,
      "Row": 6,
      "Symbol": "BackOneIcon",
      "type": "ActionKey",
    },
    Object {
      "Col": 2,
      "Label": "1",
      "Row": 0,
      "Text": "1",
      "type": "TextKey",
    },
    Object {
      "Col": 3,
      "Label": "2",
      "Row": 0,
      "Text": "2",
      "type": "TextKey",
    },
    Object {
      "Col": 9,
      "Row": 4,
      "ShiftDownLabel": "M",
      "ShiftUpLabel": "m",
      "Text": "m",
      "type": "TextKey",
    },
    Object {
      "Col": 8,
      "Row": 4,
      "ShiftDownLabel": "N",
      "ShiftUpLabel": "n",
      "Text": "n",
      "type": "TextKey",
    },
    Object {
      "Col": 11,
      "Row": 2,
      "ShiftDownLabel": "P",
      "ShiftUpLabel": "p",
      "Text": "p",
      "type": "TextKey",
    },
    Object {
      "Col": 7,
      "Row": 4,
      "ShiftDownLabel": "B",
      "ShiftUpLabel": "b",
      "Text": "b",
      "type": "TextKey",
    },
    Object {
      "Col": 10,
      "Row": 3,
      "ShiftDownLabel": "K",
      "ShiftUpLabel": "k",
      "Text": "k",
      "type": "TextKey",
    },
    Object {
      "Col": 6,
      "Row": 4,
      "ShiftDownLabel": "V",
      "ShiftUpLabel": "v",
      "Text": "v",
      "type": "TextKey",
    },
    Object {
      "Col": 0,
      "Label": "-",
      "Row": 3,
      "Text": "-",
      "type": "TextKey",
    },
    Object {
      "Col": 4,
      "Label": "3",
      "Row": 0,
      "Text": "3",
      "type": "TextKey",
    },
    Object {
      "Col": 6,
      "Label": "_",
      "Row": 1,
      "Text": "_",
      "type": "TextKey",
    },
    Object {
      "Col": 2,
      "Label": ";",
      "Row": 6,
      "Text": ";",
      "type": "TextKey",
    },
    Object {
      "Col": 0,
      "Label": "@",
      "Row": 1,
      "Text": "@",
      "type": "TextKey",
    },
    Object {
      "Col": 7,
      "Label": "\\"",
      "Row": 1,
      "Text": "\\"",
      "type": "TextKey",
    },
    Object {
      "Col": 5,
      "Label": "'",
      "Row": 1,
      "Text": "'",
      "type": "TextKey",
    },
    Object {
      "Col": 11,
      "Label": "#",
      "Row": 4,
      "Text": "#",
      "type": "TextKey",
    },
    Object {
      "Col": 1,
      "Label": "0",
      "Row": 0,
      "Text": "0",
      "type": "TextKey",
    },
    Object {
      "Col": 11,
      "Row": 3,
      "ShiftDownLabel": "L",
      "ShiftUpLabel": "l",
      "Text": "l",
      "type": "TextKey",
    },
    Object {
      "Col": 4,
      "Row": 3,
      "ShiftDownLabel": "S",
      "ShiftUpLabel": "s",
      "Text": "s",
      "type": "TextKey",
    },
    Object {
      "Col": 5,
      "Row": 2,
      "ShiftDownLabel": "R",
      "ShiftUpLabel": "r",
      "Text": "r",
      "type": "TextKey",
    },
    Object {
      "Col": 6,
      "Row": 2,
      "ShiftDownLabel": "T",
      "ShiftUpLabel": "t",
      "Text": "t",
      "type": "TextKey",
    },
    Object {
      "Col": 5,
      "Row": 4,
      "ShiftDownLabel": "C",
      "ShiftUpLabel": "c",
      "Text": "c",
      "type": "TextKey",
    },
    Object {
      "Col": 5,
      "Row": 3,
      "ShiftDownLabel": "D",
      "ShiftUpLabel": "d",
      "Text": "d",
      "type": "TextKey",
    },
    Object {
      "Col": 3,
      "Row": 2,
      "ShiftDownLabel": "W",
      "ShiftUpLabel": "w",
      "Text": "w",
      "type": "TextKey",
    },
    Object {
      "Col": 4,
      "Label": "+",
      "Row": 1,
      "Text": "+",
      "type": "TextKey",
    },
    Object {
      "Col": 1,
      "Label": "<",
      "Row": 4,
      "Text": "<",
      "type": "TextKey",
    },
    Object {
      "Col": 2,
      "Label": "=",
      "Row": 4,
      "Text": "=",
      "type": "TextKey",
    },
    Object {
      "Col": 3,
      "Row": 3,
      "ShiftDownLabel": "A",
      "ShiftUpLabel": "a",
      "Text": "a",
      "type": "TextKey",
    },
    Object {
      "Col": 4,
      "Row": 2,
      "ShiftDownLabel": "E",
      "ShiftUpLabel": "e",
      "Text": "e",
      "type": "TextKey",
    },
    Object {
      "Col": 9,
      "Row": 2,
      "ShiftDownLabel": "I",
      "ShiftUpLabel": "i",
      "Text": "i",
      "type": "TextKey",
    },
    Object {
      "Col": 10,
      "Row": 2,
      "ShiftDownLabel": "O",
      "ShiftUpLabel": "o",
      "Text": "o",
      "type": "TextKey",
    },
    Object {
      "Col": 8,
      "Row": 2,
      "ShiftDownLabel": "U",
      "ShiftUpLabel": "u",
      "Text": "u",
      "type": "TextKey",
    },
    Object {
      "Col": 4,
      "Row": 4,
      "ShiftDownLabel": "X",
      "ShiftUpLabel": "x",
      "Text": "x",
      "type": "TextKey",
    },
    Object {
      "Col": 3,
      "Row": 4,
      "ShiftDownLabel": "Y",
      "ShiftUpLabel": "y",
      "Text": "y",
      "type": "TextKey",
    },
    Object {
      "Col": 7,
      "Row": 2,
      "ShiftDownLabel": "Z",
      "ShiftUpLabel": "z",
      "Text": "z",
      "type": "TextKey",
    },
    Object {
      "Action": "LeftWin",
      "Col": 0,
      "Label": "Win",
      "Row": 5,
      "type": "ActionKey",
    },
    Object {
      "Col": 1,
      "Label": ">",
      "Row": 5,
      "Text": ">",
      "type": "TextKey",
    },
    Object {
      "Col": 2,
      "Label": ",",
      "Row": 5,
      "Text": ",",
      "type": "TextKey",
    },
    Object {
      "Col": 3,
      "Label": ".",
      "Row": 5,
      "Text": ".",
      "type": "TextKey",
    },
    Object {
      "Col": 4,
      "Label": "(",
      "Row": 5,
      "Text": "(",
      "type": "TextKey",
    },
    Object {
      "Col": 5,
      "Label": ")",
      "Row": 5,
      "Text": ")",
      "type": "TextKey",
    },
    Object {
      "Col": 6,
      "Label": "[",
      "Row": 5,
      "Text": "[",
      "type": "TextKey",
    },
    Object {
      "Col": 7,
      "Label": "]",
      "Row": 5,
      "Text": "]",
      "type": "TextKey",
    },
    Object {
      "Col": 8,
      "Label": "{",
      "Row": 5,
      "Text": "{",
      "type": "TextKey",
    },
    Object {
      "Col": 9,
      "Label": "}",
      "Row": 5,
      "Text": "}",
      "type": "TextKey",
    },
    Object {
      "Action": "ArrowUp",
      "Col": 10,
      "Row": 5,
      "Symbol": "UpArrowKeyIcon",
      "type": "ActionKey",
    },
    Object {
      "Action": "Escape",
      "Col": 11,
      "Label": "Esc",
      "Row": 5,
      "type": "ActionKey",
    },
    Object {
      "Action": "LeftCtrl",
      "Col": 0,
      "Label": "Ctrl",
      "Row": 6,
      "type": "ActionKey",
    },
    Object {
      "Action": "LeftAlt",
      "Col": 1,
      "Label": "Alt",
      "Row": 6,
      "type": "ActionKey",
    },
    Object {
      "Action": "LeftShift",
      "Col": 3,
      "Row": 6,
      "Symbol": "ShiftIcon",
      "type": "ActionKey",
    },
    Object {
      "Col": 4,
      "Row": 6,
      "Symbol": "TabIcon",
      "type": "TextKey",
    },
    Object {
      "Col": 5,
      "Row": 6,
      "Symbol": "SpaceIcon",
      "type": "TextKey",
    },
    Object {
      "Col": 7,
      "Row": 6,
      "Symbol": "EnterIcon",
      "type": "TextKey",
    },
    Object {
      "Action": "ArrowLeft",
      "Col": 9,
      "Row": 6,
      "Symbol": "LeftArrowKeyIcon",
      "type": "ActionKey",
    },
    Object {
      "Action": "ArrowDown",
      "Col": 10,
      "Row": 6,
      "Symbol": "DownArrowKeyIcon",
      "type": "ActionKey",
    },
    Object {
      "Action": "ArrowRight",
      "Col": 11,
      "Row": 6,
      "Symbol": "RightArrowKeyIcon",
      "type": "ActionKey",
    },
  ],
  "Name": "MMB2s Keyboard",
  "SymbolMargin": 1,
}
`;

exports[`READ FILES: no keyboard imported yet 1`] = `undefined`;
